vars/standardDeployJob.groovy

```diff
@@ -0,0 +1,71 @@
+def call(Map params) {
+    pipeline {
+        agent {
+            label "master"
+        }
+        parameters {
```

Comment:
How do these parameters interact with the `params` parameter on line 1? Does this section actually create the parameter input in the client Jenkins job, or do you need to set them up manually as well? It seems surprising for Jenkins to allow you to configure parameters in a function, but Jenkins pipelines do some surprising things. 🙂

Reply:
It is surprising, I didn't think it would let this either. It creates the parameter input for the client Jenkins job. The parameters section is there for the Jenkins job to still allow manual deployment to stages other than intg. The ecsDeploy job is called like this by the project deploy job:

```
standardDeployJob(
    IMAGE_NAME: "docker-image-name",
    TO_DEPLOY: params.TO_DEPLOY,
    ecsServiceName: "ecs-service-name",
    STAGE: params.STAGE,
    TEST_DELAY_SECONDS: 300
)
```
The multibranch pipeline job then calls as follows:

```
...
build(
                job: "deploy-job-name",
                parameters: [
                    string(name: "STAGE", value: "intg"),
                    string(name: "TO_DEPLOY", value: versionTag)
                ],
                wait: false)
...
```

This configuration then still allows us to retain the current setup of manually deploying to the higher environments via a separate deploy job.

I've set up a test pipeline https://jenkins.tdr-management.nationalarchives.gov.uk/job/tk-test-job and multibranch pipeline https://jenkins.tdr-management.nationalarchives.gov.uk/job/tk-test-pipeline/job/TDR-251-trigger-e2e-test-intg-deploy/ to  test this all fits together.

Reply:
That's interesting, thanks.

Minor thing, but the mix of casing in the parameter keys looks a bit confusing:

```
standardDeployJob(
    IMAGE_NAME: "docker-image-name",
    TO_DEPLOY: params.TO_DEPLOY,
    ecsServiceName: "ecs-service-name",
    STAGE: params.STAGE,
    TEST_DELAY_SECONDS: 300
)
```

Is it meaningful, or can they all be camelCase, which I think is the Groovy naming convention?

Reply:
No the casing is consistent, the example snippet is out of date. Its all AAAA_AAAA format

Reply:
The casing is consistent, just used an old code snippet. They are all in the AAA_AAA format which is the format we are using for Jenkins params generally. So this is consistent with other Jenkins jobs. To change to camel case will mean the parameters in the Jenkins job would then be inconsistent with other Jenkins jobs, ie camel case.

Reply:
Build params are unavailable inside the function. If the param is not passed in directly into the function for example with no "TO_DEPLOY" param:

```
ecsDeployJob(
    IMAGE_NAME: "docker-image-name",
    ECS_SERVICE: "ecs-service-name",
    STAGE: params.STAGE,
    TEST_DELAY_SECONDS: 300
)
```

Then there is no value available. The value set in the Parameter section in the client Jenkins job is not picked up. Thus the param has to be passed into the function directly for it to work, ie:

```
ecsDeployJob(
    IMAGE_NAME: "docker-image-name",
    **TO_DEPLOY: "some value"**
    ECS_SERVICE: "ecs-service-name",
    STAGE: params.STAGE,
    TEST_DELAY_SECONDS: 300
)
```
Can always rename the Map params something else to make a clearer distinction, if this would make it less confusing. For example  `ecsDeployJob(Map **config**)`

Reply:
I think renaming it would be useful to avoid the confusion I had about shadowed parameters. Config sounds good. 👍